README.md

<p align="center">

<img src="https://github.com/homebridge/branding/raw/master/logos/homebridge-wordmark-logo-vertical.png" width="150">

</p>


# Homebridge Plugin for Breezart Vents

homebridge-breezart is a plugin for homebridge which allows you to control your [Breezart Vents](http://breezart.ru/) from your Home app. It should work with all Breezart Vents that has controllers like **JL204С5M**, **JL205**, **JL206** and a remote **TPD-283U-H**.

For remote control using this module you must be able to connect the Breezart remote **TPD-283U-H** to LAN, activate the remote control in it and set the password.

**Attention**: Remotes must have firmware version `Vent 5.08`

## Features
* View vents status
* Power management
* Fan speed control
* Temperature control
* View vents work mode
* Energy consumption history
* Temperature history

> **Disclaimer**\
> Work in progress\
> Management features coming soon

## Installation
If you are new to homebridge, please first read the homebridge [documentation](https://www.npmjs.com/package/homebridge).

### Install homebridge
```
npm install -g homebridge
```
### Install homebridge-breezart
```
npm install -g homebridge-breezart
```

## Configuration
Add the breezart platform in config.json inside `~/.homebridge`.

Add your Breezart Vent or multiply vents in the devices or `devices` array.

Example configuration:
```json
{
  "bridge": {
    "name": "Homebridge",
    "username": "CC:22:3D:E3:CE:44",
    "manufacturer": "homebridge.io",
    "model": "homebridge",
    "port": 51853,
    "pin": "022-74-141"
  },
  "platforms": [
    {
      "platform": "Breezart",
      "devices": [
        {
          "name": "Breezart Home",
          "host": "192.168.0.20",
          "port": 1560,
          "password": 12345
        }
      ]
    }
  ],
  "accessories": []
}
```
Also you must be able to connect the Breezart remote **TPD-283U-H** to LAN, activate the remote control in it and set the password.

A history of temperature and energy consumption is available in the Elgato Eve app.


CHANGELOG.md
Changelog
=========

# 1.x release

## v1.3.1
- Fix: added missing dependency

## v1.3.0
- Enhance: added the power consumption characteristics
- Enhance: added the power consumption history
- Enhance: added the temperature history
- Other: added the version of the compatible firmware

## v1.2.0

- Enhance: added temperature control
- Enhance: added power management
- Fix: filter change indication

## v1.1.0

- Enhance: added fan speed control
- Other: added workflow for publishing to NPM

## v1.0.1

- Major: initial public release
- Enhance: added read only Fan accessory
- Enhance: added read only Heater Cooler accessory
- Enhance: added read only Filter Maintenance accessory



src/platform.ts
import { API, DynamicPlatformPlugin, Logger, PlatformAccessory, Service, Characteristic } from 'homebridge';

import { PLATFORM_NAME, PLUGIN_NAME } from './settings';
import { BreezartPlatformAccessory } from './platformAccessory';
import { BreezartPlatformConfig } from './configTypes';
import { BreezartDeviceConfig } from 'breezart-client';
import fakegato from 'fakegato-history';


/**
 * BreezartHomebridgePlatform
 * This class is the main constructor for Breezart plugin, this is where we
 * parse the user config and discover/register accessories with Homebridge.
 */
export class BreezartHomebridgePlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  // this is used to track restored cached accessories
  public readonly accessories: PlatformAccessory[] = [];

  // this is configs of the devices by uuid
  public readonly deviceConfigs: Map<string, BreezartDeviceConfig> = new Map();

  // Elgato Eve history service
  public readonly FakeGatoHistoryService;

  constructor(
    public readonly log: Logger,
    public readonly config: BreezartPlatformConfig,
    public readonly api: API,
  ) {

    this.log.debug('Finished initializing platform:', PLATFORM_NAME);

    // When this event is fired it means Homebridge has restored all cached accessories from disk.
    // Dynamic Platform plugins should only register new accessories after this event was fired,
    // in order to ensure they weren't added to homebridge already. This event can also be used
    // to start discovery of new accessories.
    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback');
      // run the method to discover / register your devices as accessories
      this.discoverDevices();
    });

    // Add Elgato Eve history service
    this.FakeGatoHistoryService = fakegato(this.api);
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);

    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.accessories.push(accessory);
  }

  /**
   * Register discovered accessories.
   * Accessories must only be registered once, previously created accessories
   * must not be registered again to prevent "duplicate UUID" errors.
   */
  discoverDevices() {

    // loop over the configured devices and register each one if it has not already been registered
    if (this.config.devices) {
      for (const device of this.config.devices) {
        if (!device.name) {
          this.log.error('One of your Breezart devices has no name configured. This device will be skipped.');
          continue;
        }
        if (!device.host) {
          this.log.error('One of your Breezart devices has no host configured. This device will be skipped.');
          continue;
        }
        if (!device.password) {
          this.log.error('One of your Breezart devices has no password configured. This device will be skipped.');
          continue;
        }
        // generate a unique id for the accessory this should be generated from
        // something globally unique, but constant, for example, the device serial
        // number or MAC address
        const uuid = this.api.hap.uuid.generate(device.host + String(device.port));

        // check the name for uniqueness and add a config of the device to config map
        if (this.deviceConfigs.has(uuid)) {
          this.log.warn('Multiple devices are configured with same host and port. Duplicates will be skipped.', device.host, device.port);
          continue;
        } else {
          this.deviceConfigs.set(uuid, device);
        }
        // see if an accessory with the same uuid has already been registered and restored from
        // the cached devices we stored in the `configureAccessory` method above
        const existingAccessory = this.accessories.find(accessory => accessory.UUID === uuid);

        if (existingAccessory) {
          // the accessory already exists
          if (device) {
            this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);

            // if you need to update the accessory.context then you should run `api.updatePlatformAccessories`. eg.:
            // existingAccessory.context.device = device;
            // this.api.updatePlatformAccessories([existingAccessory]);

            // create the accessory handler for the restored accessory
            // this is imported from `platformAccessory.ts`
            new BreezartPlatformAccessory(this, existingAccessory);
            
            // update accessory cache with any changes to the accessory details and information
            this.api.updatePlatformAccessories([existingAccessory]);
          } else if (!device) {
            // it is possible to remove platform accessories at any time using `api.unregisterPlatformAccessories`, eg.:
            // remove platform accessories when no longer present
            this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [existingAccessory]);
            this.log.info('Removing existing accessory from cache:', existingAccessory.displayName);
          }
        } else {
          // the accessory does not yet exist, so we need to create it
          this.log.info('Adding new accessory:', device.name);

          // create a new accessory
          const accessory = new this.api.platformAccessory(device.name, uuid);

          // store a copy of the device object in the `accessory.context`
          // the `context` property can be used to store any data about the accessory you may need
          accessory.context.device = device;

          // create the accessory handler for the newly create accessory
          new BreezartPlatformAccessory(this, accessory);

          // link the accessory to your platform
          this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
        }
      }
    }
  }
}


src/breezartController.ts
// import { BreezartDeviceConfig } from './configTypes';
import { BreezartClient, BreezartDeviceConfig, BreezartCallback } from 'breezart-client';


/**
 * Time that does not update the property after setting it via the setter
 */
const UPDATE_TIMEOUT = 3000;

export class BreezartController extends BreezartClient {

  private _active: boolean;
  private _rotationSpeed: number;
  private _currentTemperature: number | null;
  private _mode: number; // mode = 4 - heating/cooling is off
  private _currentHeaterCoolerState: number;
  private _heatingThresholdTemperature: number;
  private _filterChange: number; // 0|1
  private _filterLifeLevel: number; // 4%
  private _error: Error | null; // the last error from a device
  private _totalPwr: number; // 4%
  private _resetTotalPwr: number; // Set to seconds from 1.1.2001 upon Reset of total consumption in Eve.app

  private _times: Map<string, Date>;

  constructor(options: BreezartDeviceConfig) {
    super(options);
    this._times = new Map();

    this._active = false;
    this._rotationSpeed = 40;
    this._currentTemperature = 24.4;
    this._mode = 1;
    this._currentHeaterCoolerState = 0;
    this._heatingThresholdTemperature = 24;
    this._filterChange = 0;
    this._filterLifeLevel = 4;
    this._error = null;
    this._totalPwr = 0;
    this._resetTotalPwr = 0;

  }

  public get active() {
    return this._active;
  }

  public set active(value: boolean) {
    this._active = value;
    this._times['active'] = Date.now();
  }

  public get rotationSpeed() {
    return this._rotationSpeed;
  }

  public set rotationSpeed(value: number) {
    this._rotationSpeed = value;
    this._times['rotationSpeed'] = Date.now();
  }

  public get currentTemperature() {
    return this._currentTemperature;
  }

  public get mode() {
    return this._mode;
  }

  public set mode(value: number) {
    this._mode = value;
    this._times['mode'] = Date.now();
  }

  public get currentHeaterCoolerState() {
    return this._currentHeaterCoolerState;
  }

  public get heatingThresholdTemperature() {
    return this._heatingThresholdTemperature;
  }

  public set heatingThresholdTemperature(value: number) {
    this._heatingThresholdTemperature = value;
    this._times['heatingThresholdTemperature'] = Date.now();
  }

  public get filterChange() {
    return this._filterChange;
  }

  public get filterLifeLevel() {
    return this._filterLifeLevel;
  }

  public get error() {
    return this._error;
  }

  public set error(value: Error | null) {
    this._error = value;
  }

  public get totalPwr() {
    return this._totalPwr;
  }

  public set totalPwr(value: number) {
    this._totalPwr = value;
  }

  public get resetTotalPwr() {
    return this._resetTotalPwr;
  }

  public set resetTotalPwr(value: number) {
    this._resetTotalPwr = value;
  }

  PullStatus (callback: BreezartCallback) {
    this.getCurrentStatus((error) => {
      if (error) {
        callback(error);
      } else {
        // reset error, if status was received
        this._error = null;
        const expired = Date.now() - UPDATE_TIMEOUT;

        if (!('active' in this._times) || expired > this._times['active']) {
          this._active = !!this.PwrBtnState;
        }
        if (!('rotationSpeed' in this._times) || expired > this._times['rotationSpeed']) {
          this._rotationSpeed = this.SpeedTarget;
        }
        if (!('heatingThresholdTemperature' in this._times) || expired > this._times['heatingThresholdTemperature']) {
          this._heatingThresholdTemperature = this.TemperTarget;
        }
        if (!('mode' in this._times) || expired > this._times['mode']) {
          this._mode = this.ModeSet;
        }

        this._currentTemperature = this.TInf;
        this._currentHeaterCoolerState = this.Mode;
        this._filterChange = this.ChangeFilter;
        this._filterLifeLevel = this.FilterDust;

        callback();
      }
    });
  }

}


src/customCharacteristics.ts
import { Characteristic, Formats, Perms } from 'hap-nodejs';

export class CurrentPowerConsumption extends Characteristic {
  static readonly UUID: string = 'E863F10D-079E-48FF-8F27-9C2605A29F52'; // Eve Consumption

  constructor () {
    super('Consumption', CurrentPowerConsumption.UUID, {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore: custom unit
      unit: 'W',
      minStep: 0.1,
      format: Formats.FLOAT,
      perms: [Perms.PAIRED_READ, Perms.NOTIFY],
    });
  }
}

export class TotalPowerConsumption extends Characteristic {
  static readonly UUID: string = 'E863F10C-079E-48FF-8F27-9C2605A29F52'; // Eve Total Consumption

  constructor () {
    super('Total Consumption', TotalPowerConsumption.UUID, {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore: custom unit
      unit: 'kWh',
      minStep: 0.01,
      format: Formats.FLOAT,
      perms: [Perms.PAIRED_READ, Perms.NOTIFY],
    });
  }
}

export class ResetTotalPowerConsumption extends Characteristic {
  static readonly UUID: string = 'E863F112-079E-48FF-8F27-9C2605A29F52'; // Eve Reset Total Consumption

  constructor () {
    super('Reset Total Consumption', ResetTotalPowerConsumption.UUID, {
      format: Formats.UINT32,
      perms: [Perms.PAIRED_READ, Perms.NOTIFY, Perms.PAIRED_WRITE],
    });
  }
}


src/platformAccessory.ts
/* eslint-disable max-len */
import { Service, PlatformAccessory, CharacteristicValue, CharacteristicSetCallback, CharacteristicGetCallback } from 'homebridge';

import { BreezartHomebridgePlatform } from './platform';
import { BreezartDeviceConfig, BreezartEventTypes } from 'breezart-client';
import { BreezartController } from './breezartController';
import { CurrentPowerConsumption, TotalPowerConsumption, ResetTotalPowerConsumption } from './customCharacteristics';

import { uuid } from 'hap-nodejs';

/**
 * Device polling interval in mc
 */
const POLLS_INTERVAL = 1000;

/**
 * Breezart Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
export class BreezartPlatformAccessory {
  private service: Service;
  private hQService: Service;
  private fService: Service;

  private breezart: BreezartController;
  private loggingService;

  constructor(
    private readonly platform: BreezartHomebridgePlatform,
    private readonly accessory: PlatformAccessory,
  ) {

    // set accessory information
    this.accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Breezart')
      .setCharacteristic(this.platform.Characteristic.Model, 'TPD-283U-H')
      .setCharacteristic(this.platform.Characteristic.SerialNumber, 'Default-Serial');
    
    // get the Fanv2 service if it exists, otherwise create a new Fanv2 service
    this.service = this.accessory.getService(this.platform.Service.Fanv2) ||
      this.accessory.addService(this.platform.Service.Fanv2);
    // set the service name, this is what is displayed as the default name on the Home app
    this.service.setCharacteristic(this.platform.Characteristic.Name, accessory.context.device.name);
    // register handlers for the Active state
    this.service.getCharacteristic(this.platform.Characteristic.Active)
      .on('set', this.setActive.bind(this))
      .on('get', this.getActive.bind(this));
    // register handlers for the Rotation Speed
    this.service.getCharacteristic(this.platform.Characteristic.RotationSpeed)
      .on('set', this.setRotationSpeed.bind(this))
      .on('get', this.getRotationSpeed.bind(this));

    // Register Heater Cooler service
    // register the Current Temperature
    this.hQService = this.accessory.getService(this.platform.Service.HeaterCooler) ||
      this.accessory.addService(this.platform.Service.HeaterCooler, accessory.context.device.name + ' Mode');
    // register handlers for the Active state (Breezart Mode)
    this.hQService.getCharacteristic(this.platform.Characteristic.Active)
      .on('set', this.setModeActive.bind(this))
      .on('get', this.getModeActive.bind(this));
    this.hQService.getCharacteristic(this.platform.Characteristic.CurrentTemperature)
      .on('get', this.getCurrentTemperature.bind(this));
    this.hQService.getCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState)
      .on('get', this.getCurrentHeaterCoolerState.bind(this));
    this.hQService.getCharacteristic(this.platform.Characteristic.TargetHeaterCoolerState)
      .on('set', this.setTargetHeaterCoolerState.bind(this))
      .on('get', this.getTargetHeaterCoolerState.bind(this));
    this.hQService.getCharacteristic(this.platform.Characteristic.HeatingThresholdTemperature)
      .on('set', this.setHeatingThresholdTemperature.bind(this))
      .on('get', this.getHeatingThresholdTemperature.bind(this));
    // add Power Consumption characteristics for the history service
    this.hQService.getCharacteristic(CurrentPowerConsumption)
      .on('get', this.getCurrentPowerConsumption.bind(this));
    this.hQService.getCharacteristic(TotalPowerConsumption)
      .on('get', this.getTotalPowerConsumption.bind(this));
    this.hQService.getCharacteristic(ResetTotalPowerConsumption)
      .on('set', this.setResetTotalPowerConsumption.bind(this))
      .on('get', this.getResetTotalPowerConsumption.bind(this));
    
    // Register Filter Maintaince service
    this.fService = this.accessory.getService(this.platform.Service.FilterMaintenance) ||
      this.accessory.addService(this.platform.Service.FilterMaintenance, accessory.context.device.name + ' Filter');
    this.fService.getCharacteristic(this.platform.Characteristic.FilterChangeIndication)
      .on('get', this.getFilterChangeIndication.bind(this));
    this.fService.getCharacteristic(this.platform.Characteristic.FilterLifeLevel)
      .on('get', this.getFilterLifeLevel.bind(this));
    // TODO: Reset filter indications https://developers.homebridge.io/#/characteristic/ResetFilterIndication
    
    // Register Elgato Eve history service
    this.loggingService = new this.platform.FakeGatoHistoryService('custom', accessory, {
      storage: 'fs',
      filename: `AccessoryHistory.${uuid.toShortForm(accessory.UUID)}.json`,
      disableRepeatLastData: true,
      minutes: 10, // save averages every 10 minutes
      size: 12096, // 60 / 10 * 24 * 28 * 3 - Store last 3 month
      // log: this.platform.log, // block due to spam
    });

    // Link all services to Fan
    this.service.addLinkedService(this.hQService);
    this.service.addLinkedService(this.fService);

    const connectionOptions = {
      host: accessory.context.device.host as string,
      port: accessory.context.device.port as number,
      password: accessory.context.device.password as number,
    };

    this.breezart = new BreezartController(connectionOptions as BreezartDeviceConfig);
    // Connect to Brizart device and start the polling
    this.initBreezartClient(accessory.context.device);
  }

  getActive(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	1
    const isActive = this.breezart.active;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic Active ->', isActive);
    callback(error, isActive);
  }

  setActive(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    const targetPowerState = value as boolean;
    this.breezart.active = targetPowerState;
    this.platform.log.debug('Set Characteristic Active ->', value);

    this.breezart.setPowerOn(targetPowerState, callback);
  }

  getRotationSpeed(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	100
    // Min Step	1
    const rotationSpeed = this.breezart.rotationSpeed;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic RotationSpeed ->', rotationSpeed);
    callback(error, rotationSpeed);
  }

  setRotationSpeed(value: CharacteristicValue, callback: CharacteristicSetCallback) {
  
    const targetSpeed = value as number < this.breezart.SpeedMin ? this.breezart.SpeedMin : value as number;
    this.breezart.rotationSpeed = targetSpeed;

    this.platform.log.debug('Set Characteristic RotationSpeed ->', value);

    this.breezart.setFanSpeed(targetSpeed, callback);
  }

  getCurrentTemperature(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	100
    // Min Step	0.1
    const currentTemperature = this.breezart.currentTemperature;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic CurrentTemperature ->', currentTemperature);
    callback(error, currentTemperature);
  }

  getModeActive(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	1
    let modeActive = this.platform.Characteristic.Active.ACTIVE;
    if (this.breezart.mode === 4) {
      modeActive = this.platform.Characteristic.Active.INACTIVE;
    }

    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic modeActive ->', modeActive);
    callback(error, modeActive);
  }

  setModeActive(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    if (value !== this.breezart.mode) {
      if (value === this.platform.Characteristic.Active.INACTIVE) {
        this.breezart.mode = 4;
        this.platform.log.debug('Set Characteristic modeActive ->', this.breezart.mode);
      }
    }
    const error = this.breezart.error;
    callback(error);
  }

  getCurrentHeaterCoolerState(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	3
    let currentHeaterCoolerState;
    switch (this.breezart.currentHeaterCoolerState) {
      case 0:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.HEATING;
        break;
      case 1:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.COOLING;
        break;
      case 2:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.HEATING;
        break;
      case 3:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.COOLING;
        break;
      case 4:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.IDLE;
        break;
      default:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.INACTIVE;
        break;
    }
    this.platform.log.debug('Get Characteristic currentHeaterCoolerState ->', currentHeaterCoolerState);
    const error = this.breezart.error;
    callback(error, currentHeaterCoolerState);
  }

  getTargetHeaterCoolerState(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	2
    const mode = this.breezart.mode;
    let targetHeaterCoolerState;
    switch (mode) {
      case 1:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.HEAT;
        break;
      case 2:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.COOL;
        break;
      default:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.AUTO;
        break;
    }
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic targetHeaterCoolerState ->', targetHeaterCoolerState);
    callback(error, targetHeaterCoolerState);
  }

  setTargetHeaterCoolerState(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    switch (value as number) {
      case this.platform.Characteristic.TargetHeaterCoolerState.HEAT:
        this.breezart.mode = 1;
        break;
      case this.platform.Characteristic.TargetHeaterCoolerState.COOL:
        this.breezart.mode = 2;
        break;
      default:
        this.breezart.mode = 3;
        break;
    }
    const error = this.breezart.error;
    this.platform.log.debug('Set Characteristic targetHeaterCoolerState ->', value);
    callback(error);
  }
  
  getHeatingThresholdTemperature(callback: CharacteristicGetCallback) {
    // Min Value	0
    // Max Value	25
    // Min Step	0.1
    const targetTemperature = this.breezart.heatingThresholdTemperature;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic heatingThresholdTemperature ->', targetTemperature);
    callback(error, targetTemperature);
  }
  
  setHeatingThresholdTemperature(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    const targetTemp = Math.round(value as number);
    this.breezart.heatingThresholdTemperature = targetTemp;
    this.platform.log.debug('Set Characteristic heatingThresholdTemperature ->', value);
    this.breezart.setTemperature(targetTemp, callback);
  }
  
  getCurrentPowerConsumption(callback: CharacteristicGetCallback) {
    const currentPowerConsumption = this.breezart.Pwr;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic CurrentPowerConsumption ->', currentPowerConsumption);
    callback(error, currentPowerConsumption);
  }
  
  getTotalPowerConsumption(callback: CharacteristicGetCallback) {
    const totalPowerConsumption = this.breezart.totalPwr;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic TotalPowerConsumption ->', totalPowerConsumption);
    callback(error, totalPowerConsumption);
  }
  
  getResetTotalPowerConsumption(callback: CharacteristicGetCallback) {
    const resetTotalPowerConsumption = this.breezart.resetTotalPwr;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic ResetTotalPowerConsumption ->', resetTotalPowerConsumption);
    callback(error, resetTotalPowerConsumption);
  }

  setResetTotalPowerConsumption(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    const resetTotalPowerConsumption = value as number;
    this.breezart.resetTotalPwr = resetTotalPowerConsumption;
    this.breezart.totalPwr = 0;
    const error = this.breezart.error;
    this.setExtraPersistedData();
    this.platform.log.debug('Set Characteristic resetTotalPowerConsumption ->', value);
    callback(error, resetTotalPowerConsumption);
  }
  
  getFilterChangeIndication(callback: CharacteristicGetCallback) {
    const filterChange = this.breezart.filterChange;
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic filterChange ->', filterChange);
    callback(error, filterChange);
  }

  getFilterLifeLevel(callback: CharacteristicGetCallback) {
    let filterLifeLevel;
    if (this.breezart.filterLifeLevel === 255 || this.breezart.filterLifeLevel === null) {
      // it's like undefined
      filterLifeLevel = 0;
    } else {
      filterLifeLevel = this.breezart.filterLifeLevel > 100 ? 100 : this.breezart.filterLifeLevel;
    }
    const error = this.breezart.error;
    this.platform.log.debug('Get Characteristic filterChange ->', filterLifeLevel);
    callback(error, filterLifeLevel);
  }

  /**
   * Connet to Breezart device and init polling
   */
  initBreezartClient(options: BreezartDeviceConfig) {
    this.platform.log.debug('initBreezartClient');

    // then Breezart was connected 
    this.breezart.on(BreezartEventTypes.CONNECT, () => {
      this.platform.log.debug('Connection established', options.name);
      // Update accessory information
      this.accessory.getService(this.platform.Service.AccessoryInformation)!
        .setCharacteristic(this.platform.Characteristic.FirmwareRevision, `${String(this.breezart.HiVerTPD)}.${String(this.breezart.LoVerTPD)}`)
        .setCharacteristic(this.platform.Characteristic.SerialNumber, String(this.breezart.Firmware_Ver));
      // Set props for heater/cooler
      this.hQService.getCharacteristic(this.platform.Characteristic.HeatingThresholdTemperature)
        .setProps({
          minValue: this.breezart.TempMin,
          maxValue: this.breezart.TempMax,
          minStep: 1,
        });
      // Set props for fan
      this.service.getCharacteristic(this.platform.Characteristic.RotationSpeed)
        .setProps({
          maxValue: this.breezart.SpeedMax,
          minValue: this.breezart.SpeedMin - 1,
          minStep: 1,
        });

      this.startPolls(POLLS_INTERVAL);
    });
    // if device disconnected - reconnect after 5 sec
    this.breezart.on(BreezartEventTypes.DISCONNECT, () => {
      this.platform.log.debug('Device disconnected', options.name);
      if (!this.breezart.connected) {
        setTimeout(() => {
          this.breezart.connect();
        }, 5000);
      }
    });
    // handle errors
    this.breezart.on(BreezartEventTypes.ERROR, (err: Error) => {
      // if from device was received an error, set it to device state
      this.breezart.error = err;
      this.platform.log.error(err.message);
    });

    this.breezart.connect();
  }

  /**
   * Start periodical polls of the device for reading characteristics
   */
  startPolls(interval: number) {
    setInterval(() => {
      if (!this.breezart.connected) {
        return;
      }
      this.breezart.PullStatus((error) => {
        if (error) {
          return;
        }
        this.setDeviceStates();
      });
    }, interval);
  }

  setDeviceStates() {
    // active (like the Power On)
    this.service.updateCharacteristic(this.platform.Characteristic.Active, this.breezart.active);
    // fan rotation speed
    this.service.updateCharacteristic(this.platform.Characteristic.RotationSpeed, this.breezart.rotationSpeed);
    // current temperature
    if (this.breezart.currentTemperature) {
      this.hQService.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, this.breezart.currentTemperature);
    }
    // mode
    const modeActive = this.breezart.mode === 4 ? 0 : 1;
    this.hQService.updateCharacteristic(this.platform.Characteristic.Active, modeActive);
    const mode = this.breezart.mode;
    let targetHeaterCoolerState;
    switch (mode) {
      case 1:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.HEAT;
        break;
      case 2:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.COOL;
        break;
      default:
        targetHeaterCoolerState = this.platform.Characteristic.TargetHeaterCoolerState.AUTO;
        break;
    }
    this.hQService.updateCharacteristic(this.platform.Characteristic.TargetHeaterCoolerState, targetHeaterCoolerState);

    // current heater coolerState
    let currentHeaterCoolerState;
    switch (this.breezart.currentHeaterCoolerState) {
      case 0:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.HEATING;
        break;
      case 1:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.COOLING;
        break;
      case 2:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.HEATING;
        break;
      case 3:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.COOLING;
        break;
      case 4:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.IDLE;
        break;
      default:
        currentHeaterCoolerState = this.platform.Characteristic.CurrentHeaterCoolerState.INACTIVE;
        break;
    }
    this.hQService.updateCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState, currentHeaterCoolerState);

    // heatingThresholdTemperature
    this.hQService.updateCharacteristic(this.platform.Characteristic.HeatingThresholdTemperature, this.breezart.heatingThresholdTemperature);

    // Pwr
    this.hQService.updateCharacteristic(CurrentPowerConsumption, this.breezart.Pwr ? this.breezart.Pwr : 0);

    // filterChange
    this.fService.updateCharacteristic(this.platform.Characteristic.FilterChangeIndication, this.breezart.filterChange);

    // filterLifeLevel
    let filterLifeLevel;
    if (this.breezart.filterLifeLevel === 255 || this.breezart.filterLifeLevel === null) {
      // it's like undefined
      filterLifeLevel = 0;
    } else {
      filterLifeLevel = this.breezart.filterLifeLevel > 100 ? 100 : this.breezart.filterLifeLevel;
    }
    this.fService.updateCharacteristic(this.platform.Characteristic.FilterLifeLevel, filterLifeLevel);


    // Calculate the Total Power Consumption metric
    const deltaTotalPower = (this.breezart.Pwr ? this.breezart.Pwr : 0) * (POLLS_INTERVAL / 1000) / 3600 / 1000;
    let persistedTotalPower = 0;
    let resetTotalPowerConsumption = 0;
    if (this.loggingService.isHistoryLoaded()) {
      const extraPersistedData = this.loggingService.getExtraPersistedData();
      persistedTotalPower = extraPersistedData?.totalPower || 0;
      resetTotalPowerConsumption = extraPersistedData?.resetPower || 0;
    }
    const totalPower = persistedTotalPower + deltaTotalPower;
    this.breezart.totalPwr = totalPower;
    this.breezart.resetTotalPwr = resetTotalPowerConsumption;
    this.setExtraPersistedData();

    // TotalPowerConsumption
    this.hQService.updateCharacteristic(TotalPowerConsumption, Math.round((this.breezart.totalPwr + Number.EPSILON) * 100) / 100);
    // resetTotalPowerConsumption
    this.hQService.updateCharacteristic(ResetTotalPowerConsumption, resetTotalPowerConsumption);

    // add history
    const moment = Math.round(new Date().valueOf() / 1000);
    const entryPwr = {
      time: moment,
      power: this.breezart.Pwr,
    };
    this.loggingService.addEntry(entryPwr);
    const entryTmp = {
      time: moment,
      temp: this.breezart.currentTemperature,
    };
    this.loggingService.addEntry(entryTmp);

  }

  setExtraPersistedData() {
    this.loggingService.setExtraPersistedData({
      totalPower: this.breezart.totalPwr,
      resetPower: this.breezart.resetTotalPwr,
    });
  }

}

